File: GoEngineering/09_database_repository_layer/explanation/09_database_repository_layer_explanation.go
/*
Package repository — Database Repository Layer (Problem 09)

WHAT WE'RE BUILDING

A repository layer over database/sql for a simple "accounts" domain (an
account has an ID, an owner name, and a balance in cents), backed by
modernc.org/sqlite — a pure-Go, cgo-free SQLite driver. The repository owns
connection pooling, prepared statements, context propagation, and row
scanning behind a small, testable interface. This same domain (accounts with
a balance) continues into Problem 10, which adds transactions, optimistic
locking, and concurrent-update safety on top of what's built here.

# WHY IT MATTERS IN REAL SYSTEMS

database/sql is a thin, driver-agnostic abstraction: almost every production
bug in a hand-rolled repository layer traces back to misunderstanding one of
a small number of things it does NOT do for you — it does not retry, it does
not manage transactions for you beyond what you explicitly Begin/Commit, and
critically, it pools connections in a way that has sharp edges with SQLite
specifically. Getting comfortable with *Row/*Rows scanning, prepared
statement reuse, context-aware queries (QueryContext/ExecContext, never the
context-less variants in a service), and connection pool tuning is
foundational to every service that talks to a real database.

THE `:memory:` + CONNECTION POOL FOOTGUN (read this before writing any code)

sql.Open("sqlite", ":memory:") does NOT give you one in-memory database
shared across your whole process. database/sql maintains a POOL of
connections and opens new ones lazily as needed; with modernc.org/sqlite (and
essentially every SQLite driver), each physical connection to ":memory:" is
its own separate, independent in-memory database. So the very first query
that happens to grab connection #2 from the pool sees a "database" with none
of the tables you created on connection #1 — intermittent, hard-to-reproduce
"table not found" or "no rows" bugs that appear only under concurrency (once
the pool actually opens more than one connection).

Two correct fixes:
 1. db.SetMaxOpenConns(1) — force the whole pool down to a single physical
    connection, so ":memory:" is always the same database. Correct, but
    means every query is fully serialized — fine for tests, wrong for
    anything with real concurrent load.
 2. Use a real file on disk — even a t.TempDir()-rooted temp file — as the
    DSN. Every pooled connection opens the SAME file, so this works
    correctly with the pool's default (concurrent) behavior. This is what
    this problem's tests do, and what you should default to for anything
    beyond a single quick throwaway script.

(There is a third option — SQLite's shared-cache mode via a
"file::memory:?cache=shared" DSN — which also fixes it, but has its own
locking quirks and is generally considered more surprising than just using a
temp file, so we don't use it here.)

CONCEPTS COVERED
  - database/sql: sql.DB is a pool, not a connection — Open doesn't connect
  - Connection pool tuning: SetMaxOpenConns, SetMaxIdleConns, SetConnMaxLifetime
  - Prepared statements: *sql.Stmt reuse vs ad-hoc query strings
  - Context propagation: QueryContext/ExecContext/QueryRowContext everywhere,
    never the non-Context variants, so callers can cancel/timeout DB calls
  - Row scanning: sql.ErrNoRows, scanning into typed structs, NULL handling
    via sql.NullString/sql.NullInt64 or driver-level *string
  - Resource cleanup: defer rows.Close(), checking rows.Err() after the loop
  - Schema setup for tests: golang-migrate-style raw DDL executed once per
    test DB

REQUIREMENTS / SPEC

	type Account struct {
	    ID      int64
	    Owner   string
	    Balance int64 // cents, to avoid float rounding issues with money
	}

	type Repository struct { ... } // wraps *sql.DB + prepared statements

	func Open(ctx context.Context, dsn string) (*Repository, error)
	    Opens the DB (sql.Open + Ping to fail fast on a bad DSN), configures
	    the connection pool (document your chosen limits and why), runs
	    schema migration (CREATE TABLE IF NOT EXISTS), and prepares the
	    statements this repository will reuse. Must accept ctx and propagate
	    it into the Ping and schema-setup calls.

	func (r *Repository) Close() error
	    Closes prepared statements then the DB. Must be safe to call once
	    (document whether double-Close is safe — sql.DB.Close is idempotent,
	    prepared statement Close should be handled the same way or guarded).

	func (r *Repository) CreateAccount(ctx context.Context, owner string, initialBalanceCents int64) (Account, error)
	func (r *Repository) GetAccount(ctx context.Context, id int64) (Account, error)
	    Not found -> a wrapped error satisfying errors.Is(err, ErrNotFound).
	func (r *Repository) ListAccounts(ctx context.Context, ownerLike string) ([]Account, error)
	    ownerLike "" means "all accounts"; otherwise a SQL LIKE pattern
	    against owner (caller supplies %-wildcards).
	func (r *Repository) UpdateBalance(ctx context.Context, id int64, newBalanceCents int64) error
	    Not found -> ErrNotFound. (Problem 10 replaces/extends this with
	    optimistic-locking and transactional variants — keep this one
	    simple: a straight UPDATE.)
	func (r *Repository) DeleteAccount(ctx context.Context, id int64) error
	    Not found -> ErrNotFound.

ACCEPTANCE CRITERIA
  - Every exported method takes a context.Context as its first parameter and
    uses the *Context variant of the relevant database/sql call — no bare
    Query/Exec/QueryRow anywhere in the repository.
  - GetAccount/UpdateBalance/DeleteAccount on a nonexistent ID return an
    error satisfying errors.Is(err, ErrNotFound), never a bare sql.ErrNoRows
    leaking out of the package.
  - Tests use t.TempDir() + a file-based SQLite DSN (per the footgun note
    above), never ":memory:" with the default pool settings, and each test
    gets its own fresh DB file (hermetic, safe under -parallel).
  - Close() releases all resources and a second Close() call does not panic.
  - ListAccounts with ownerLike == "" returns every account; with a LIKE
    pattern, only matches.
  - go vet and go test -race are clean.

Read the HINTS and PITFALLS at the bottom before opening the solution file.
*/
package repository

import (
	"context"
	"errors"
)

// Account is one row of the accounts table. Balance is stored in integer
// cents specifically to avoid the rounding/comparison hazards of storing
// money as a float.
type Account struct {
	ID      int64
	Owner   string
	Balance int64
}

// ErrNotFound is returned (wrapped, via fmt.Errorf("...: %w", ErrNotFound))
// by any method that looked up a row by ID and found none. Callers should
// use errors.Is(err, ErrNotFound), never compare against sql.ErrNoRows
// directly — that's a database/sql implementation detail this package must
// not leak.
var ErrNotFound = errors.New("repository: account not found")

// Repository wraps a *sql.DB (and its prepared statements) for the accounts
// domain.
type Repository struct {
	// TODO: hold *sql.DB and any *sql.Stmt you choose to prepare once and
	// reuse (e.g. the GetAccount-by-id query, hit on every balance check).
}

// Open opens dsn, configures the connection pool, ensures the schema
// exists, and returns a ready-to-use *Repository.
//
// IMPORTANT: read the ":memory:" + connection pool footgun note in the
// package doc comment above before choosing a DSN in tests or callers.
func Open(ctx context.Context, dsn string) (*Repository, error) {
	// TODO:
	//  1. sql.Open("sqlite", dsn) — this does NOT connect yet, it just
	//     validates the DSN format and sets up the pool's config struct.
	//  2. db.PingContext(ctx) — actually opens a connection, so bad DSNs /
	//     unreachable databases fail here, at Open time, not on the first
	//     real query deep inside request handling.
	//  3. Configure the pool: SetMaxOpenConns, SetMaxIdleConns,
	//     SetConnMaxLifetime. Pick real numbers and justify them in a
	//     comment (this is a SQLite file DB in tests/local dev — the
	//     numbers that make sense here differ from a Postgres pool sized
	//     for a busy service; say why).
	//  4. Run "CREATE TABLE IF NOT EXISTS accounts (...)" via
	//     ExecContext(ctx, ...) to ensure schema exists.
	//  5. Prepare any statements you want to reuse via db.PrepareContext.
	//  6. Wrap it all in a *Repository and return it.
	panic("TODO: implement Open")
}

// Close releases the repository's prepared statements and underlying
// *sql.DB. Safe to call more than once.
func (r *Repository) Close() error {
	// TODO: Close prepared statements (best-effort, join errors), then
	// db.Close(). sql.DB.Close() is documented idempotent; decide/document
	// whether your prepared-statement closing needs its own guard.
	panic("TODO: implement Close")
}

// CreateAccount inserts a new account and returns it with its assigned ID.
func (r *Repository) CreateAccount(ctx context.Context, owner string, initialBalanceCents int64) (Account, error) {
	// TODO: INSERT via ExecContext, then result.LastInsertId() for the new
	// row's ID (SQLite supports this via the rowid alias, as long as the
	// primary key is `INTEGER PRIMARY KEY`).
	panic("TODO: implement CreateAccount")
}

// GetAccount fetches one account by ID. Returns an error satisfying
// errors.Is(err, ErrNotFound) if no such account exists.
func (r *Repository) GetAccount(ctx context.Context, id int64) (Account, error) {
	// TODO: QueryRowContext + Scan; translate sql.ErrNoRows to a wrapped
	// ErrNotFound via fmt.Errorf("...: %w", ErrNotFound) (NOT
	// errors.Join — you want a single linear chain here so
	// errors.Is(err, ErrNotFound) and errors.Is(err, sql.ErrNoRows) can
	// both work if you wrap sql.ErrNoRows too, though only ErrNotFound is
	// part of this package's public contract).
	panic("TODO: implement GetAccount")
}

// ListAccounts returns every account whose owner matches ownerLike (a SQL
// LIKE pattern), or every account if ownerLike == "".
func (r *Repository) ListAccounts(ctx context.Context, ownerLike string) ([]Account, error) {
	// TODO: two query paths (all rows vs LIKE-filtered), or one query with
	// a `WHERE ? = '' OR owner LIKE ?` — pick one and justify it. Remember:
	// defer rows.Close(), and check rows.Err() after the scan loop (a
	// mid-stream I/O error can surface there even if every individual Scan
	// succeeded).
	panic("TODO: implement ListAccounts")
}

// UpdateBalance sets id's balance to newBalanceCents directly (no
// optimistic locking here — see Problem 10 for that). Returns ErrNotFound
// if id doesn't exist.
func (r *Repository) UpdateBalance(ctx context.Context, id int64, newBalanceCents int64) error {
	// TODO: ExecContext an UPDATE, then check result.RowsAffected() == 0
	// to detect "no such id" (an UPDATE matching zero rows is NOT an error
	// from database/sql's point of view — you must check RowsAffected
	// yourself).
	panic("TODO: implement UpdateBalance")
}

// DeleteAccount removes id. Returns ErrNotFound if id doesn't exist.
func (r *Repository) DeleteAccount(ctx context.Context, id int64) error {
	// TODO: same RowsAffected-based existence check as UpdateBalance.
	panic("TODO: implement DeleteAccount")
}

/*
HINTS
  - sql.Open never returns an error for "the database is unreachable" — it
    only validates the DSN's shape. The FIRST real network/file I/O happens
    lazily, on the first query OR on an explicit Ping/PingContext. Always
    Ping in Open() so misconfiguration fails at startup, not on a random
    user's first request.
  - result.RowsAffected() is how you detect "my UPDATE/DELETE matched
    nothing" — database/sql does not turn "0 rows affected" into an error.
  - sql.ErrNoRows is returned ONLY by QueryRow's .Scan(), never by Query
    (Query on zero rows just gives you a *sql.Rows that iterates zero
    times, no error).
  - Prepared statements (*sql.Stmt) obtained from *sql.DB are themselves
    pool-aware: calling stmt.QueryContext concurrently is safe and reuses
    the pool's connections. You don't need one *sql.Stmt per connection.

COMMON PITFALLS
  - Using ":memory:" as the DSN in tests without SetMaxOpenConns(1) — read
    the footgun section above. This produces flaky, pool-state-dependent
    "table not found" failures that are miserable to debug because they
    often pass locally (small pool, few queries) and fail in CI (more
    concurrent test parallelism triggering a second connection).
  - Calling Query/Exec/QueryRow (no "Context" suffix) anywhere in service
    code — these use context.Background() internally, silently defeating
    caller-supplied timeouts and cancellation.
  - Forgetting `defer rows.Close()` after Query/QueryContext — leaks the
    underlying connection back to the pool late (or never, if a panic
    happens before an explicit Close later in the function), which under
    load exhausts the pool.
  - Not checking rows.Err() after the scan loop — an error reading the next
    row from the wire/file mid-iteration is only surfaced there, not from
    Scan or the loop condition.
  - Treating "0 rows affected" from an UPDATE/DELETE as success — it means
    "nothing matched", which for an ID-keyed lookup means the ID doesn't
    exist and should surface as ErrNotFound, not silently look like it
    worked.

STRETCH GOALS
  - Add a `WithTx(ctx, func(*Repository) error) error` helper that begins a
    transaction, hands the caller a Repository bound to that tx, and
    commits/rolls back based on the callback's return — foreshadows
    Problem 10.
  - Benchmark CreateAccount with and without a prepared statement to
    measure SQLite's parse/plan overhead per call.
  - Add an index on `owner` and benchmark ListAccounts with a LIKE prefix
    query before/after, discussing why a LIKE '%foo%' pattern can't use a
    standard b-tree index even with one present.
  - Add OpenTelemetry-style span instrumentation around each method to
    show where per-call context propagation earns its keep operationally.
*/

File: PyEngineering/09_database_repository_layer/09_database_repository_layer_explanation.py
"""
09 — Database Repository Layer
=================================

WHAT
----
Build a repository layer around stdlib `sqlite3` for an "inventory" domain:

    class Product(NamedTuple / dataclass): id, sku, name, quantity, price_cents
    class ProductRepository:
        create(sku, name, quantity, price_cents) -> Product
        get(product_id) -> Product | None
        get_by_sku(sku) -> Product | None
        list_all() -> list[Product]
        update_quantity(product_id, quantity) -> None
        delete(product_id) -> None
        search_by_name(query: str) -> list[Product]   # deliberately shows
                                                          the injection risk

WHY THIS MATTERS
----------------
`sqlite3` is the stdlib's only bundled database, but it is *not* toy-grade —
it's the actual embedded database inside SQLite-backed production systems
(mobile apps, browsers, and plenty of backend services for embedded/edge
deployments). The engineering concerns here — connection lifetime,
thread-safety, injection-safe querying, and row-to-object mapping — transfer
directly to any DB-API-2.0-shaped driver (psycopg2, mysqlclient) you'll use
against Postgres/MySQL in a real service. Problem 10 builds transactions and
concurrency control directly on top of this repository.

CONNECTION HANDLING
--------------------
`sqlite3.connect(path)` returns a connection that, by default, has
`check_same_thread=True`: using it from a thread other than the one that
created it raises `ProgrammingError`. There are two legitimate real-world
patterns:
  1. **One connection per thread** (e.g. a thread-local connection pool) —
     the most robust, avoids any cross-thread sharing entirely.
  2. **A single shared connection with `check_same_thread=False`** plus your
     own external locking — only safe if you fully understand SQLite's own
     locking model underneath (a whole connection-level lock per writer);
     this is what problem 10 explores deliberately, including the
     `database is locked` failure mode.
This module keeps it simple: `ProductRepository` owns exactly one
connection for its lifetime, used as a context manager so the connection is
always closed. Multi-connection/thread-safety concerns are pushed to
problem 10 where they're the actual subject.

THE INJECTION RISK (do this deliberately, once, to feel it)
--------------------------------------------------------------
`search_by_name` must be implemented TWICE for this exercise:
  - `search_by_name_unsafe(query)` — builds SQL via an f-string. Demonstrate
    (in the test file) that `query = "' OR '1'='1"` returns every row,
    proving the vulnerability exists, not just asserting it in prose.
  - `search_by_name(query)` — the real, shipped method — uses a parameterized
    query (`?` placeholders + a tuple of bound values) so the DB driver
    handles escaping; the same malicious `query` string returns zero rows
    (or only rows that literally contain that substring), because it's
    treated as *data*, never as *SQL*.

ROW MAPPING
------------
Configure the connection's `row_factory = sqlite3.Row` so query results
support both index and column-name access (`row["name"]`, `row[1]`), then
map each `sqlite3.Row` to a `Product` dataclass at the repository boundary —
callers of `ProductRepository` should never see a raw `sqlite3.Row` or
tuple; the DB library's return shape is an implementation detail that must
not leak past this layer.

SPEC
----
    @dataclass(frozen=True)
    class Product:
        id: int
        sku: str
        name: str
        quantity: int
        price_cents: int

    class ProductNotFoundError(Exception): ...
    class DuplicateSkuError(Exception): ...

    class ProductRepository:
        def __init__(self, db_path: str | Path) -> None: ...
            # opens the connection, creates the table if missing
        def close(self) -> None: ...
        def __enter__(self) -> "ProductRepository": ...
        def __exit__(self, *exc_info: object) -> None: ...

        def create(self, sku: str, name: str, quantity: int, price_cents: int) -> Product: ...
            # raises DuplicateSkuError on a UNIQUE constraint violation
        def get(self, product_id: int) -> Product: ...
            # raises ProductNotFoundError if missing
        def get_by_sku(self, sku: str) -> Product | None: ...
        def list_all(self) -> list[Product]: ...
        def update_quantity(self, product_id: int, quantity: int) -> None: ...
            # raises ProductNotFoundError if missing
        def delete(self, product_id: int) -> None: ...
            # raises ProductNotFoundError if missing
        def search_by_name(self, query: str) -> list[Product]: ...
            # parameterized LIKE query, injection-safe
        def search_by_name_unsafe(self, query: str) -> list[Product]: ...
            # deliberately vulnerable, for the injection-demo test only

ACCEPTANCE CRITERIA
--------------------
- `sqlite3.Row` (or any raw tuple/row) never crosses the repository's public
  API boundary — every public method returns `Product` (or `None`/raises).
- Creating a product with a duplicate `sku` raises `DuplicateSkuError`, not
  a raw `sqlite3.IntegrityError`.
- Every query with user-controllable input uses `?` placeholders — the test
  suite explicitly demonstrates `search_by_name_unsafe`'s vulnerability and
  `search_by_name`'s immunity to the identical crafted input.
- `get`/`update_quantity`/`delete` on a missing id raise
  `ProductNotFoundError`, not `None` or a silent no-op — silent no-ops on a
  missing row hide real bugs (e.g. deleting an already-deleted order twice
  should be loud, not silently "successful").
- The connection is always closed (context manager `__exit__`/explicit
  `close()`), even when an exception propagates out of a `with` block.
- Table creation is idempotent (`CREATE TABLE IF NOT EXISTS`) so opening the
  same `db_path` twice in a row never raises.
- Tests use `tmp_path / "test.db"` for a fresh file-backed SQLite DB per
  test (real SQLite, not mocked) — this exercises actual constraint
  enforcement, actual type affinity, actual locking, which a mock cannot.
"""

from __future__ import annotations

import sqlite3
from dataclasses import dataclass
from pathlib import Path


@dataclass(frozen=True)
class Product:
    id: int
    sku: str
    name: str
    quantity: int
    price_cents: int


class ProductNotFoundError(Exception):
    """Raised when a lookup/update/delete targets a missing product id."""


class DuplicateSkuError(Exception):
    """Raised when creating a product whose sku already exists."""


class ProductRepository:
    """Repository over a single sqlite3 connection for the inventory domain."""

    def __init__(self, db_path: str | Path) -> None:
        """TODO:
        - open sqlite3.connect(str(db_path))
        - set connection.row_factory = sqlite3.Row
        - CREATE TABLE IF NOT EXISTS products (
              id INTEGER PRIMARY KEY AUTOINCREMENT,
              sku TEXT NOT NULL UNIQUE,
              name TEXT NOT NULL,
              quantity INTEGER NOT NULL,
              price_cents INTEGER NOT NULL
          )
        - commit the DDL
        """
        raise NotImplementedError

    def close(self) -> None:
        """TODO: close the connection."""
        raise NotImplementedError

    def __enter__(self) -> ProductRepository:
        raise NotImplementedError

    def __exit__(self, *exc_info: object) -> None:
        raise NotImplementedError

    def _row_to_product(self, row: sqlite3.Row) -> Product:
        """TODO: map a sqlite3.Row to a Product. Keep this the ONLY place
        that knows about column names/order."""
        raise NotImplementedError

    def create(self, sku: str, name: str, quantity: int, price_cents: int) -> Product:
        """TODO: INSERT with ? placeholders. Catch sqlite3.IntegrityError
        from the UNIQUE constraint and re-raise as DuplicateSkuError. Use
        cursor.lastrowid to fetch and return the created Product."""
        raise NotImplementedError

    def get(self, product_id: int) -> Product:
        """TODO: SELECT by id with a ? placeholder. Raise
        ProductNotFoundError if no row."""
        raise NotImplementedError

    def get_by_sku(self, sku: str) -> Product | None:
        """TODO: SELECT by sku with a ? placeholder. Return None if
        missing (this one's optional-by-design, unlike get())."""
        raise NotImplementedError

    def list_all(self) -> list[Product]:
        """TODO: SELECT all rows ordered by id."""
        raise NotImplementedError

    def update_quantity(self, product_id: int, quantity: int) -> None:
        """TODO: UPDATE with ? placeholders. Check cursor.rowcount == 0 to
        detect a missing id and raise ProductNotFoundError (rowcount is
        SQLite's way of telling you whether the WHERE clause matched
        anything, without a separate SELECT)."""
        raise NotImplementedError

    def delete(self, product_id: int) -> None:
        """TODO: DELETE with a ? placeholder; rowcount == 0 ->
        ProductNotFoundError."""
        raise NotImplementedError

    def search_by_name(self, query: str) -> list[Product]:
        """TODO: parameterized LIKE search: WHERE name LIKE ? with bound
        value f"%{query}%" passed as a *parameter*, never interpolated into
        the SQL string itself."""
        raise NotImplementedError

    def search_by_name_unsafe(self, query: str) -> list[Product]:
        """TODO (deliberately vulnerable, for the injection-demo test ONLY):
        build the SQL via an f-string/`.format()` with `query` spliced
        directly into the WHERE clause. Never write real code like this —
        this method exists solely so the test suite can prove the
        vulnerability exists and that search_by_name avoids it."""
        raise NotImplementedError


# ---------------------------------------------------------------------------
# Hints
# ---------------------------------------------------------------------------
# - `cursor.execute("... WHERE id = ?", (product_id,))` — always pass a
#   tuple/sequence of parameters, even for a single value (the trailing
#   comma matters: `(x,)` is a 1-tuple, `(x)` is just `x`).
# - `cursor.lastrowid` after an INSERT gives you the AUTOINCREMENT id
#   without a second round-trip query.
# - `cursor.rowcount` after an UPDATE/DELETE tells you how many rows
#   matched — 0 means "no such id", the correct/cheap way to detect a
#   missing row without doing a SELECT first (which would also race).
# - Wrap the UNIQUE-constraint INSERT in `try: ... except sqlite3.
#   IntegrityError as exc: raise DuplicateSkuError(...) from exc`.
#
# Pitfalls
# --------
# - Forgetting `row_factory = sqlite3.Row` and returning raw index-only
#   tuples from internal helpers — fragile the moment column order changes.
# - Using `%s`-style or f-string interpolation for *any* user-controllable
#   value in SQL — sqlite3 uses `?` placeholders (not `%s`, that's the
#   psycopg2/MySQL convention) or named `:name` placeholders.
# - Not committing DDL/DML — sqlite3's default `isolation_level` starts an
#   implicit transaction on the first DML statement; forgetting `commit()`
#   silently loses writes when the connection closes uncleanly. (Problem 10
#   goes deep on isolation levels.)
# - Returning `None` instead of raising for a missing id on `delete`/
#   `update_quantity` — makes "nothing happened" indistinguishable from
#   "it worked", which hides real bugs like double-deletes.
#
# Stretch goals
# -------------
# - Add `bulk_create(items: list[tuple]) -> None` using
#   `cursor.executemany` and measure/comment on why it's faster than a
#   Python-level loop of individual `execute` calls.
# - Add an index on `sku` explicitly (even though UNIQUE already implies
#   one) and explain in a comment when you'd want a *non-unique* index vs.
#   relying on the implicit one.
# - Add `.pragma_foreign_keys(True)` scaffolding in preparation for a
#   second table (e.g. `orders` referencing `products`) in a later problem.
